### PR TITLE
HW #14 "Concurrency"

### PR DESCRIPTION
**Pull request by sviat-h**

### CODE REVIEW

File: src/com/cursor/hw14/task1/controller/AppExecutor.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package com.cursor.hw14.task1.controller;
+
+import com.cursor.hw14.task1.service.UserRead;
+import com.cursor.hw14.task1.service.UserWrite;
+
+public class AppExecutor {
+    public void runProgram() {
+        UserWrite userWrite = new UserWrite();
```
Comment (omilovskiy) [inline src/com/cursor/hw14/task1/controller/AppExecutor.java]: add empty line above

File: src/com/cursor/hw14/task1/database/DB.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        if (instance == null) {
+            instance = new DB();
+        }
+        return instance;
+    }
+
+    public void addInfo(Integer key, String value) {
+        synchronized (dataBase) {
+            System.out.println(
+                    "Thread with id " + Thread.currentThread().getId() + " got access to dataBase and write data.");
```
Comment (omilovskiy) [inline src/com/cursor/hw14/task1/database/DB.java]: as previous 
and fix formatting here

File: src/com/cursor/hw14/task1/database/DB.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            dataBase.put(key, value);
+
+            dataBase.notifyAll();
+        }
+    }
+
+    public String getInfo(Integer key, long timeToWaitIfNoValueAvailable) {
+            System.out.println(
+                    "Thread with id " + Thread.currentThread().getId() + " got access to dataBase to read data.");
```
Comment (omilovskiy) [inline src/com/cursor/hw14/task1/database/DB.java]: the same

File: src/com/cursor/hw14/task1/database/DB.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            final Date deadLine = new Date(System.currentTimeMillis() + timeToWaitIfNoValueAvailable);
```
Comment (omilovskiy) [inline src/com/cursor/hw14/task1/database/DB.java]: Don’t forget about empty lines

File: src/com/cursor/hw14/task1/database/DB.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+            while (!dataBase.containsKey(key) && new Date().before(deadLine)) {
```
Comment (omilovskiy) [inline src/com/cursor/hw14/task1/database/DB.java]: new Date() in loop??

File: src/com/cursor/hw14/task1/database/DB.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+                try {
+                    Thread.sleep(timeToWaitIfNoValueAvailable);
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
+            }
+            if (!dataBase.containsKey(key)) {
+                throw new IllegalArgumentException(String.format("No object with key %s found in database", key));
+            }
```
Comment (omilovskiy) [inline src/com/cursor/hw14/task1/database/DB.java]: fix formatting here
hard to read

File: src/com/cursor/hw14/task1/service/UserRead.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        joiner.add(database.getInfo(3, TIME_TO_WAIT_FOR_RESULT));
+        joiner.add(database.getInfo(4, TIME_TO_WAIT_FOR_RESULT));
+        joiner.add(database.getInfo(5, TIME_TO_WAIT_FOR_RESULT));
+        joiner.add(database.getInfo(6, TIME_TO_WAIT_FOR_RESULT));
+        joiner.add(database.getInfo(7, TIME_TO_WAIT_FOR_RESULT));
+        try {
+            Thread.sleep(400);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
```
Comment (omilovskiy) [inline src/com/cursor/hw14/task1/service/UserRead.java]: create some common method with a parameter for this reason

File: src/com/cursor/hw14/task1/service/UserRead.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        joiner.add(database.getInfo(13, TIME_TO_WAIT_FOR_RESULT));
+        joiner.add(database.getInfo(14, TIME_TO_WAIT_FOR_RESULT));
+        joiner.add(database.getInfo(15, TIME_TO_WAIT_FOR_RESULT));
+        joiner.add(database.getInfo(16, TIME_TO_WAIT_FOR_RESULT));
+        joiner.add(database.getInfo(17, TIME_TO_WAIT_FOR_RESULT));
+        try {
+            Thread.sleep(400);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
```
Comment (omilovskiy) [inline src/com/cursor/hw14/task1/service/UserRead.java]: the same

File: src/com/cursor/hw14/task1/service/UserWrite.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        database.addInfo(0, "info 0");
+        database.addInfo(1, "info 1");
+        database.addInfo(2, "info 2");
+        database.addInfo(3, "info 3");
+        database.addInfo(4, "info 4");
+        try {
+            Thread.sleep(100);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
```
Comment (omilovskiy) [inline src/com/cursor/hw14/task1/service/UserWrite.java]: also

File: src/com/cursor/hw14/task1/service/UserWrite.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+        database.addInfo(7, "info 7");
+        database.addInfo(8, "info 8");
+        database.addInfo(9, "info 9");
+        database.addInfo(10, "info 10");
+        database.addInfo(11, "info 11");
+        try {
+            Thread.sleep(400);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
```
Comment (omilovskiy) [inline src/com/cursor/hw14/task1/service/UserWrite.java]: yep

File: src/com/cursor/hw14/task1/controller/AppExecutor.java (hunk 1 above)
Comment (sviat-h) [reply src/com/cursor/hw14/task1/controller/AppExecutor.java]: +

File: src/com/cursor/hw14/task1/database/DB.java (hunk 2 above)
Comment (sviat-h) [reply src/com/cursor/hw14/task1/database/DB.java]: +

File: src/com/cursor/hw14/task1/database/DB.java (hunk 3 above)
Comment (sviat-h) [reply src/com/cursor/hw14/task1/database/DB.java]: +

File: src/com/cursor/hw14/task1/database/DB.java (hunk 4 above)
Comment (sviat-h) [reply src/com/cursor/hw14/task1/database/DB.java]: +

File: src/com/cursor/hw14/task1/database/DB.java (hunk 5 above)
Comment (sviat-h) [reply src/com/cursor/hw14/task1/database/DB.java]: +

File: src/com/cursor/hw14/task1/database/DB.java (hunk 6 above)
Comment (sviat-h) [reply src/com/cursor/hw14/task1/database/DB.java]: +

File: src/com/cursor/hw14/task1/service/UserRead.java (hunk 7 above)
Comment (sviat-h) [reply src/com/cursor/hw14/task1/service/UserRead.java]: +

File: src/com/cursor/hw14/task1/service/UserRead.java (hunk 8 above)
Comment (sviat-h) [reply src/com/cursor/hw14/task1/service/UserRead.java]: +

File: src/com/cursor/hw14/task1/service/UserWrite.java (hunk 9 above)
Comment (sviat-h) [reply src/com/cursor/hw14/task1/service/UserWrite.java]: +

File: src/com/cursor/hw14/task1/service/UserWrite.java (hunk 10 above)
Comment (sviat-h) [reply src/com/cursor/hw14/task1/service/UserWrite.java]: +

**Pull request merged**